#!/bin/bash
#SBATCH --job-name=gbrs_one_slurm
#SBATCH -p compute # partition (queue)
#SBATCH -q batch 
#SBATCH -N 1 # number of nodes
#SBATCH -n 1 # number of cores
#SBATCH --mem 512GB # memory pool for all cores
#SBATCH -t 0-12:00 # time (D-HH:MM)
#SBATCH --mail-type=ALL
#SBATCH	--mail-user=samantha.ardery@jax.org 
# example use: sbatch --export=ALL,SAMPLE="PB360_26",SEX='F',GENERATION='16',INDIR="/projects/munger-lab/raw/DO_mESC/Read1",OUTDIR="/projects/munger-lab/SampleReruns/SE_Data" /home/arders/SampleRerun/Scripts/SE_SampleRerun.slurm

# check that variables SAMPLE, INDIR and OUTDIR exist
if [ -z "$SAMPLE" ]; then echo "SAMPLE is unset"; exit 1; fi # e.g. DO-1290
if [ -z "$SEX" ]; then echo "SEX is unset"; exit 1; fi # M or F
if [ -z "$GENERATION" ]; then echo "GENERATION is unset"; exit 1; fi # e.g. G8
if [ -z "$INDIR" ]; then echo "INDIR is unset"; exit 1; fi # FASTQ.GZ files folder
if [ -z "$OUTDIR" ]; then echo "OUTDIR is unset"; exit 1; fi # output folder


module load singularity
cd $OUTDIR
export GBRS_DATA=/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/hmm

echo "SE Sample Rerun"

echo "0) Merging FASTQ files and setting variables"
echo `date`

#TEMPDIR=`mktemp -d`
TEMPDIR="/fastscratch/arders/temp"
echo "Files in Temporal Directory ${TEMPDIR}"
echo "Sample $SAMPLE"

# merge gzipped fastq files into one FASTQ file
SEQFILE="${TEMPDIR}/${SAMPLE}.fastaq"
zcat `ls ${INDIR}/${SAMPLE}*` > ${SEQFILE}

# 1) bowtie options
INDEXBASE="/projects/munger-lab/Genomes/bowtie1/transcripts"
OUTFILE1a="$TEMPDIR/$SAMPLE.8-way.transcriptome.sam"
OUTFILE1b="$TEMPDIR/$SAMPLE.8-way.transcriptome.bam"

# 2) gbrs bam2emase options
LIDFILE="/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/ref/emase.transcripts.info"
SUFFICES=A:B:C:D:E:F:G:H
OUTFILE2="$TEMPDIR/$SAMPLE.8-way.transcriptome.h5"
INDEX_DTYPE=uint32
DATA_DTYPE=uint8

# 3) gbrs compress options
OUTFILE2c="$TEMPDIR/$SAMPLE.8-way.transcriptome.compressed.h5"

# 4) gbrs quantify_multiway options - quantifying multiway allele specificity
GRPFILE="/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/ref/emase.gene2transcripts.tsv"
LENFILE="/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/8-way/emase.pooled.transcripts.info"
MODEL=4
OUTFILEBASE="$TEMPDIR/$SAMPLE"
#READLEN=100
#PSEUDOCOUNT=0.0
#MAXITERS=999
#TOLERANCE=0.0001

# Note the output files are:
# $OUTFILEBASE.multiway.genes.expected_read_counts
# $OUTFILEBASE.multiway.genes.tpm
# and the two corresponding *.isoforms.* files

# 5) gbrs reconstruct options
EXPRFILE="$OUTFILEBASE.multiway.genes.tpm"
AVECFILE="/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/hmm/avecs.npz"
TPROBFILE="/projects/churchill-lab/resource/Custom_Genomes/R84-REL1505/hmm/tranprob.DO.G${GENERATION}.${SEX}.npz"
GPOSFILE="/projects/munger-lab/gbrs/ref.gene_pos.ordered.npz"
EXPRCUTOFF=1.5
SIGMA=0.12

# What are the output files?
# .genoprobs.npz ?
# .genotypes.tsv

# 6) gbrs quantify2 options
GENOFILE="$OUTFILEBASE.genotypes.tsv"
OUTFILEBASE2="$TEMPDIR/$SAMPLE"

echo "1) Bowtie alignment"
echo `date`

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/bowtie:v1.1.2 -q -a --best --strata --sam -v 3 ${INDEXBASE} ${SEQFILE} > ${OUTFILE1a}

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/samtools:v0.1.18 view -bS -F 4 ${OUTFILE1a} > ${OUTFILE1b}

echo "2) BAM to EMASE conversion"
echo `date`

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/gbrs:v0.1.6 bam2emase -i ${OUTFILE1b} -m ${LIDFILE} -s ${SUFFICES//\:/\,} -o ${OUTFILE2} 


echo "3) Compress"
echo `date`

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/gbrs:v0.1.6 compress -i ${OUTFILE2} -o ${OUTFILE2c}

echo "4) Quantify multiway"
echo `date`

#run-emase -i ${OUTFILE2} -g ${GRPFILE} -M ${MODEL} -o ${OUTFILE3} -r ${READLEN} -p ${PSEUDOCOUNT} -m ${MAXITERS} -t ${TOLERANCE} -L ${LENFILE}
singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/gbrs:v0.1.6 quantify -i ${OUTFILE2c} -g ${GRPFILE} -L ${LENFILE} -M ${MODEL} -o ${OUTFILEBASE} --report-alignment-counts

echo "5) Reconstruct"
echo `date`

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/gbrs:v0.1.6 reconstruct -e ${EXPRFILE} -t ${TPROBFILE} -x ${AVECFILE} -g ${GPOSFILE} -c ${EXPRCUTOFF} -s ${SIGMA} -o ${OUTFILEBASE}

echo "6) Quantify"
echo `date`

singularity run library://jaxreg.jax.org/rna-seq-do-gbrs/gbrs:v0.1.6 quantify -i ${OUTFILE2c} -G ${GENOFILE} -g ${GRPFILE} -L ${LENFILE} -M ${MODEL} -o ${OUTFILEBASE2} --report-alignment-counts

# echo "7) Interpolate"
# echo `date`

# singularity run /home/aydins/sif/gbrs_0.1.6.sif gbrs interpolate -i ${GPROBFILE} -g ${GRIDFILE} -p ${GPOSFILE} -o ${OUTFILE4}

echo "8) MOVE RESULTS FROM TEMP DIRECTORY TO OUTPUT DIRECTORY"
echo `date`

cp "$TEMPDIR/$SAMPLE.8-way.transcriptome.bam" "$OUTDIR"
cp "$TEMPDIR/$SAMPLE.8-way.transcriptome.sam" "$OUTDIR" 


cp "$TEMPDIR/$SAMPLE.8-way.transcriptome.h5" "$OUTDIR"
cp "$TEMPDIR/$SAMPLE.multiway.genes.expected_read_counts" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.multiway.genes.tpm" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.multiway.isoforms.expected_read_counts" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.multiway.isoforms.tpm" "$OUTDIR"

cp "$TEMPDIR/$SAMPLEdiploid.genes.expected_read_counts" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.diploid.genes.tpm" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.diploid.isoforms.expected_read_counts" "$OUTDIR" 
cp "$TEMPDIR/$SAMPLE.diploid.isoforms.tpm" "$OUTDIR"

cp "$TEMPDIR/$SAMPLE.genotypes.tsv" "$OUTDIR"
cp "$TEMPDIR/$SAMPLE.genoprobs.npz" "$OUTDIR"
#cp "$TEMPDIR/$SAMPLE.gbrs.interpolated.genoprobs.npz" "$OUTDIR"

rm "${TEMPDIR}/$SAMPLE.*"
